Tasks/DotNetCoreInstallerV0/dotnetcoreinstaller.ts

```diff
@@ -133,22 +200,51 @@ class DotnetCoreInstaller {
 
         // cache tool
         console.log(tl.loc("CachingTool"));
-        let cachedDir = await toolLib.cacheDir(extPath, this.cachedToolName, this.version, this.arch);
-        console.log(tl.loc("SuccessfullyInstalled", this.packageType, this.version));
+        let cachedDir: string;
+        if (globalJson) {            
+            let availableGlobalTool = toolLib.findLocalTool(this.cachedToolName, "0.0.0", this.arch);
+            if (availableGlobalTool == null || availableGlobalTool.length == 0) {
```

Comment:
No caching does not look good. Can you build a simple caching logic here based on directory name? This coupled with the comment below to fail on error will ensure that the task is fast and accurate. Otherwise, it will differ in behavior from non-global-json scenario.

Reply:
If it okay when I check only the directories. Because I haven't found a good solution to check if the cached objects are valid.

Reply:
I now have build a simple check [here](https://github.com/Microsoft/azure-pipelines-tasks/pull/9301/commits/a2185c64c005e1f12668ea81e717eaa9b7763c04#diff-d0badcc911e7be9cfb54b03863b65c09R55).

Reply:
@bishal-pdMSFT please give feedback if this is okay

Reply:
@paule96 I sincerely apologize for delay. Got busy with couple of other stuffs.

I am not sure checking for one dll is correct/sufficient. In my opinion we should check just for folder existence. If version folder exists, we should trust that all files are there.

However, this trust will only build if during copy we do not set continue on error to true. I get your logic to put it because of error during overwrite, but we should think how we can work around this. Continue on error hides real error which causes trust deficit.

One option which I am thinking is to create a list of version folders which we need to copy. If all these folders exist, we trust that the version is installed. If one or more do not exist, we delete the existing version folders first and then copy each folder separately. This will ensure that copies happen on a clean slate. And here we can turn off continueOnError. The question here is how do we detect list of version folders. One option is hard-code the list - something like 
- "host\fxr",
- "sdk\",
- "shared\Microsoft.NETCore.App"
etc

Other option is to iterate through all files in extract folder and find out unique paths which contain version string regex. This logic has been used in .Net Core's official install [ps script](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script). Check out the function named `Get-List-Of-Directories-And-Versions-To-Unpack-From-Dotnet-Package`

Once we know the list of folders which need to written, we have high confidence that we can copy them properly, catch any real error and build the trust on version folder's existence.

Let me know what do you think. or have any other suggestions.

Reply:
Hi @bishal-pdMSFT,

thanks for your feedback and your idea.

The behavior for cache detection is now:

- check all known folders of this sdk version
- if on of the folders don't exsists load the sdk again

To avoid that we don't know if the copy isn't successfull the replace behavior is now:

- remove old files
- create new files

So this is not a real replace. It's more a cleanup and a new creation.

Reply:
Update after found out the caching was broken:

The behavior for cache detection is now:

- check only the sdk folder
- if the folder don't exsists load the sdk again

This is because the versioning of the sdk is different from the libary versioning. For example:

- sdk: 2.2.102
- host/fxr: 2.2.1

To avoid that we don't know if the copy isn't successfull the replace behavior is now:

- remove old files
- create new files

Reply:
@paule96 I was discussing this with @sachinma and he showed me a simple change where if we only try to overwrite the directories (i.e. exclude dotnet.exe and license file etc), then tl.cp does not complain while over-writing. If this is true, then the logic can become much simpler. We just overwrite the all directories and then create a markup file.
@sachinma can you please paste the code for reference.

Reply:
I have not thoroughly tested it. I just tried it on windows and it seemed to work for me. if extPath is where your extracted the new version, and toolRootPath is where you want to copy, then following does the trick of copying with overwrite.

```
for (let itemName of fs.readdirSync(extPath)) {
                    let s = path.join(extPath, itemName);
                    if (tl.stats(s).isDirectory()) {
                        tl.cp(s, this.toolPathRoot + '/', '-rf');
                    }
                }
```

Reply:
But why. I don't get it.

you check if this is a dictionary and then you copy the content. But what is with the single files? And what about subdictionarys that already exsists? This errors too? Or I'm wrong?

Reply:
I was able to copy over subdirectories as well. Only single files which are skipped are dotnet.exe and license etc. For Dotnet.exe, probably we should copy the highest version and that is why I was trying to not force overwrite it every time. To overwrite it, we can try to look at the existing versions present in the cache and update it only if the current version being copies is higher than that.

Also, I haven't done extensive testing. I tried it for a couple of times - and I was able to (force) overwrite sub-dirs without any issue. The failure you saw could be a timing issue and just that I didn't encounter it. 

Sub-directories would need over-writing only when the last attempt to copy failed for certain reason, and hence we do not have completed marker file. In that case, failure during copy should result in the task failure.

Reply:
hi @sachinma and @bishal-pdMSFT,

I have now another version. This is a combination if the implementation from @sachinma  and @paule96 .

So the behavior is now:
1. check if the item to copy is a folder
   1. If it true:
      1. Iterate over all items and check if they present in the destination path
      1. If it not present copy it
   1. if it false:
      1. we know that are the dotnet.exe, the license and the third party notices.
      1. so we check the version number and replace this files if they are newer.

with this method, we must only know the folder "sdk" in a dotnet installation zip. The rest of the installation is generic. hope this solve all our problems :)

> attention: if the copy fails after creating the file, it will never be replacing it....

Reply:
I am not sure if 1.ia will be safe; the reason is in case of error, we might miss copying all the files.

I would recommend, after copying a version successfully, create a marker file. On rerunning with the same version, we will skip if the marker file already exists.
And if the marker file is not present, then we should try to force overwrite. If there are errors, let us fail the task. Note that the failure will only result if the 1st attempt copied half the files; and the 2nd attempt also fails to overwrite.

Reply:
No because we check the source. Not the destination.

so 1. check the source folder. So if the Item in the result of `let itemName of fs.readdirSync(extPath)` is a folder, then it itterate over all files. So a "normal" copy error is okay and will be handelt.

Only a error while coping a single file will be result in a unstable state. Some of the error can be:
- hard shutdown of the agent while copying
- drive fail while copying

Reply:
but you are right maybe we need a metafile too for the installations. But this is not a discussion on this comment.